File: BasicsWithJava/ArtBookFragments/app/src/main/java/com/sumere/artbookfragments/view/ListFragment.java
package com.sumere.artbookfragments.view;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.sumere.artbookfragments.adapter.ListAdapter;

import com.sumere.artbookfragments.R;
import com.sumere.artbookfragments.databinding.FragmentListBinding;
import com.sumere.artbookfragments.model.Art;
import com.sumere.artbookfragments.roomdb.ArtDao;
import com.sumere.artbookfragments.roomdb.ArtDatabase;

import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;


public class ListFragment extends Fragment {

    private ListAdapter listAdapter;
    private FragmentListBinding binding;
    private ArtDatabase artDatabase;
    private ArtDao artDao;
    private final CompositeDisposable mDisposable = new CompositeDisposable();

    public ListFragment(){

    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        artDatabase = Room.databaseBuilder(requireContext(),
                        ArtDatabase.class, "Arts")
                .build();

        artDao = artDatabase.artDao();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentListBinding.inflate(inflater, container, false);
        View view = binding.getRoot();
        return view;
    }
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext());
        binding.recyclerView.setLayoutManager(layoutManager);
        getData();
    }
    public void getData() {
        mDisposable.add(artDao.getArtWithNameAndId()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(ListFragment.this::handleResponse));

    }
    private void handleResponse(List<Art> artList) {

        binding.recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        listAdapter = new ListAdapter(artList);
        binding.recyclerView.setAdapter(listAdapter);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
        mDisposable.clear();
    }
}

File: KotlinCryptoCrazy/app/src/main/java/com/sumere/kotlincryptocrazy/view/ListFragment.kt
package com.sumere.kotlincryptocrazy.view

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.sumere.kotlincryptocrazy.databinding.FragmentListBinding
import com.sumere.kotlincryptocrazy.model.CryptoModel
import com.sumere.kotlincryptocrazy.service.CryptoAPI
import com.sumere.kotlincryptocrazy.viewmodel.CryptoViewModel
import org.koin.android.ext.android.get
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel

class ListFragment : Fragment(), RecyclerViewAdapter.Listener {

    private var _binding: FragmentListBinding? = null
    private val binding get() = _binding!!
    private var cryptoAdapter = RecyclerViewAdapter(arrayListOf(),this)
    private val viewModel by viewModel<CryptoViewModel>()

    //private val api = get<CryptoAPI>()
    //private val apiLazy by inject<CryptoAPI>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        _binding = FragmentListBinding.inflate(inflater,container,false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val layout = LinearLayoutManager(requireContext())
        binding.recyclerView.layoutManager = layout
        //viewModel = ViewModelProvider(this).get(CryptoViewModel::class.java)
        viewModel.getDataFromAPI()
        observeLiveData()
    }

    private fun observeLiveData(){
        viewModel.cryptoList.observe(viewLifecycleOwner,Observer{ cryptos->
            binding.recyclerView.visibility = View.VISIBLE
            cryptoAdapter = RecyclerViewAdapter(ArrayList(cryptos.data ?: arrayListOf()),this)
            binding.recyclerView.adapter = cryptoAdapter
        })
        viewModel.cryptoError.observe(viewLifecycleOwner, Observer {
            if(it.data == true){
                binding.recyclerView.visibility = View.GONE
                binding.cryptoErrorText.visibility = View.VISIBLE
                binding.cryptoProgressBar.visibility = View.GONE
            } else{
                binding.cryptoErrorText.visibility = View.GONE
            }
        })
        viewModel.cryptoLoading.observe(viewLifecycleOwner, Observer {
            if(it.data == true){
                binding.recyclerView.visibility = View.GONE
                binding.cryptoErrorText.visibility = View.GONE
                binding.cryptoProgressBar.visibility = View.VISIBLE
            } else{
                binding.cryptoProgressBar.visibility = View.GONE
            }
        })
    }


    override fun onItemClick(cryptoModel: CryptoModel) {
        Toast.makeText(requireContext(),"Clicked On: ${cryptoModel.currency}",Toast.LENGTH_LONG).show()
    }

}